spec/implementation/HashToG2.cry

```diff
@@ -0,0 +1,176 @@
+  Y = Fp_2.mul (Fp_2.mul (Fp_2.sq Z, xd) , yn)
+
+map_fp2_times_Zz : {m1,m2,n} (fin n, fin m2, m1 >= n, m2 >= n) => [m1]t_Fp_2 -> [m2]t_Fp_2 -> [n]t_Fp_2
+map_fp2_times_Zz ks zzs = zipWith (curry Fp_2.mul) (take`{front=n}ks) (drop`{back=n}zzs)
+
+map_fp2 kzzs x = last rs where
+  rs = [last kzzs] # [F.add (F.mul (ri, x), kzzi) | kzzi <- tail (reverse kzzs) | ri <- rs]
+
+Zz_powers : t_Fp_2 -> [3]t_Fp_2
+Zz_powers Z = reverse [Fp_2.sq Z, Fp_2.sq (Fp_2.sq Z), Fp_2.mul (Fp_2.sq (Fp_2.sq Z), Fp_2.sq Z)]
+
+add_n_dbl : {n} (fin n, n >= 1) =>
+            ProjectivePoint t_Fp_2 ->
+            ProjectivePoint t_Fp_2 ->
+            ProjectivePoint t_Fp_2
+add_n_dbl out p = doublings ! 0 where
+  outp = point_dadd Fp_2 out p Fp_2.field_zero True
+  doublings = [outp] # [point_double Fp_2 d | d <- doublings | i <- [1 .. n]]
+
+times_minus_z : ProjectivePoint t_Fp_2 -> ProjectivePoint t_Fp_2
+times_minus_z P = out6 where
+  out1 = point_double Fp_2 P
+  out2 = add_n_dbl`{2} out1 P
+  out3 = add_n_dbl`{3} out2 P
+  out4 = add_n_dbl`{9} out3 P
+  out5 = add_n_dbl`{32} out4 P
+  out6 = add_n_dbl`{16} out5 P
+
+psi_psi_double_impl: ProjectivePoint t_Fp_2 -> ProjectivePoint t_Fp_2
+psi_psi_double_impl P = psi_imp (psi_imp (point_double Fp_2 P))
+
+clear_cofactor_E2_impl:  ProjectivePoint t_Fp_2 -> ProjectivePoint t_Fp_2
+clear_cofactor_E2_impl P = out4 where
+  out1 = psi_psi_double_impl P
+  t0_1 = point_neg Fp_2 P
+  t1   = psi_imp t0_1
+  out2 = point_dadd Fp_2 out1 t0_1 Fp_2.field_zero True
+  out3 = point_dadd Fp_2 out2 t1 Fp_2.field_zero True
+  t0_2 = point_dadd Fp_2 (times_minus_z P) P Fp_2.field_zero True
+  t0_3 = point_dadd Fp_2 t0_2 t1 Fp_2.field_zero True
+  t1'  = times_minus_z t0_3
+  out4 = point_dadd Fp_2 out3 t1' Fp_2.field_zero True
+
+// A slightly simplified version that we could use in a proof that the implementation corresponds to the specification:
+_map_to_isogenous_E2_impl : t_Fp_2 -> ProjectivePoint t_Fp_2
+_map_to_isogenous_E2_impl u = res where
+  // x1n, x2n, and xd:
+  Zuu = Fp_2.mul (Z, Fp_2.sq u)
+  tv2 = Fp_2.add (Fp_2.sq Zuu, Zuu)
+  x1n = Fp_2.mul (Fp_2.add (tv2, Fp_2.field_unit), B')
+  x2n = Fp_2.mul (Zuu, x1n)
+  xd' = Fp_2.mul (Fp_2.neg A', tv2)
+  xd = if F_is_zero Fp_2 xd' then Fp_2.mul(Z,A') else xd'
+  // y1:
+  gxd = Fp_2.mul(xd, Fp_2.sq xd) // xd^3
+  gx1 = F_sum Fp_2 [Fp_2.mul(x1n, Fp_2.sq x1n), F_prod Fp_2 [A', x1n, Fp_2.sq xd], Fp_2.mul(B', gxd)]
+  tv4 = F_prod Fp_2 [gx1, gxd, Fp_2.sq gxd] // gx1 gxd^3
+  e2 = is_square tv4
+  xn = if e2 then x1n else x2n
+  y' = if e2
+        then F_prod Fp_2 [gx1, gxd, Fp_2.div (sqrt tv4, tv4)]
+        else F_prod Fp_2 [sqrt (Fp_2.div(ZZZ, tv4)), gx1, gxd, Fp_2.sq u, u]
+  y = if sgn0(u) == sgn0(y') then y' else Fp_2.neg y'
+  // convert to Jacobian coordinates
+  res = (Fp_2.mul (xn, xd), Fp_2.mul (y, gxd), xd)
+
+// others facts we would need for the proof
+
+_one_of_is_square u = Fp_2.is_val u /\ ~(is_square x) ==> is_square b where
```

Comment:
A simpler assumption is possible, I think:

    Fp_2.is_val x /\ ~(is_square x) ==> is_square (Fp_2.mul (x, recip_ZZZ))

which is true (in any field) because ZZZ is not a square, so dividing by it turns squares into nonsquares and vice versa.  But rather than changing the property you could just add a comment.

Reply:
thanks, added a comment